flips/20211007-transaction-fees.md

```diff
@@ -0,0 +1,327 @@
+
+The effort cost parameters would need be adjusted infrequently. The reasons for adjusting them would be:
+
+1. After the calculation of variable transaction fees was changed.
+
+2. Over time, due to technological factors, the same amount of effort will be easier to handle by the nodes.
+
+### Inclusion effort
+
+Inclusion effort can represent many different parts of the system. Listed below are some potential factors that may play into the inclusion effort part of the fees listed by node type. Adding factors to this list (and actually implementing them) can happen gradually. Not all of these factors play equally into the total inclusion effort and it is possible that some factors are completely negligible.
+
+1. Access node:
+    1. Sending a transaction causes load on the access node (might depend on the byte size of the transaction).
+    2. The access node needs to potentially do a check, that the payer can pay (at a constant effort).
+    3. The access node should check the transaction code for syntactical correctness (most likely correlated to the transactions size)
+    4. ...
+2. Collection node:
+    1. Load because of tx size (might depend on the byte size of the transaction).
+    2. Include tx in a collection (at a constant effort).
+    3. ...
+3. Execution node:
+    1. check signatures are valid (constant x number of signatures)
+    2. ...
+4. Verification node:
+    1. ...
+
+A way to notice additional factors that might play into the inclusion effort is to compare transactions that have similar inclusion effort but cause a different load on the system.
+
+For the first iteration the inclusion effort could be defined to be a linear function of the byte length of the (rlp encoded) transaction (scaled to match average execution effort).
+
+### Execution effort
+
+The execution effort represents the effort needed to execute the transaction script (the CPU time) and the effort to handle the execution results (transferring the necessary data to the verification node for the transaction to be verified). Because execution of the transactions script depends on the current execution state, it is impossible to definitively predict the execution effort before the actual on-chain execution of the transaction.
+
+Currently the execution effort is calculated by adding 1 for every loop or function call. This could be sufficient for the first implementation. In the future the calculation of the execution effort would be changed so that function calls would have a different execution effort cost assigned to them according to the time complexity of the method.
+
+Thi model should be replaced so that the execution effort better correlates with the time needed to execute the transaction. A possible replacement is to make some functions (that take longer to run) use more effort than others.
```

Comment:
```suggestion
This model should be replaced so that the execution effort better correlates with the time needed to execute the transaction. A possible replacement is to make some functions (that take longer to run) use more effort than others.
```